x/superfluid/keeper/migrate.go

```diff
@@ -0,0 +1,103 @@
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+	}
+
+	// Check that lockID corresponds to sender, and contains correct denomination of LP shares.
+	lock, err := k.validateLockForUnpool(ctx, sender, poolIdLeaving, lockId)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+	}
+	gammSharesInLock := lock.Coins[0]
+	preUnlockLock := *lock
+
+	// Before we break the lock, we must note the time remaining on the lock.
+	// We will be freezing the concentrated liquidity position for this duration.
+	freezeDuration := k.getExistingLockRemainingDuration(ctx, lock)
+
+	// If superfluid delegated, superfluid undelegate
+	// This also burns the underlying synthetic osmo
+	err = k.unbondSuperfluidIfExists(ctx, sender, lockId)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+	}
+
+	// Finish unlocking directly for locked locks
+	// this also unlocks locks that were in the unlocking queue
+	err = k.lk.ForceUnlock(ctx, *lock)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+	}
+
+	// If shares to migrate is not specified, we migrate all shares.
+	if sharesToMigrate.IsZero() {
+		sharesToMigrate = gammSharesInLock
+	}
+
+	// Otherwise, we must ensure that the shares to migrate is less than or equal to the shares in the lock.
+	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
+	if sharesToMigrate.Amount.GT(gammSharesInLock.Amount) {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
+	}
+
+	// Exit the balancer pool position.
+	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+	}
+	// Defense in depth, ensuring we are returning exactly two coins.
+	if len(exitCoins) != 2 {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, fmt.Errorf("Balancer pool must have exactly two tokens")
+	}
+
+	frozenUntil = ctx.BlockTime().Add(freezeDuration)
+
+	// Create a full range (min to max tick) concentrated liquidity position.
+	amount0, amount1, liquidity, err = k.clk.CreateFullRangePosition(ctx, concentratedPool, sender, exitCoins, frozenUntil)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+	}
+
+	// If there are remaining gamm shares, we must re-lock them.
+	if !remainingGammShares.IsZero() {
+		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), freezeDuration)
+		newLockId = newLock.ID
+		if err != nil {
+			return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, time.Time{}, err
+		}
+		// If the lock was unlocking, we begin the unlock from where it left off.
+		if preUnlockLock.IsUnlocking() {
+			_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
```

Comment:
nit: we can call `beginUnlock` directly

Reply:
I don't think we can, unless we want to export beginUnlock from the lockup module.